File: jddata/app/src/main/java/com/example/jddata/service/BaseAction.java
package com.example.jddata.service;

public class BaseAction {
    public static final String SEARCH = "search";   // 搜索
    public static final String SEARCH_AND_SHOP = "search_and_shop";   // 搜索加购
    public static final String CART = "cart";       // 购物车
    public static final String HOME = "home";       // 首页
    public static final String BRAND_KILL = "brand_kill";   // 品牌秒杀
    public static final String BRAND_KILL_AND_SHOP = "brand_kill_and_shop";   // 品牌秒杀加购
    public static final String TYPE_KILL = "type_kill";     // 品类秒杀
    public static final String LEADERBOARD = "leaderboard";     // 排行榜
    public static final String JD_KILL = "jd_kill";       // 京东秒杀
    public static final String WORTH_BUY = "worth_buy";       // 发现好货
    public static final String NICE_BUY = "nice_buy";       // 会买专辑
    public static final String DMP = "dmp";       // DMP广告
    public static final String DMP_AND_SHOP = "dmp_and_shop";       // DMP广告加购


    public String actionType;
    public ActionMachine machine;

    public BaseAction(String actionType) {
        this.actionType = actionType;
        this.machine = new ActionMachine(actionType);
    }
}

File: jddata/app/src/main/java/com/example/jddata/service/BaseAction.kt
package com.example.jddata.service

open class BaseAction(var actionType: String) {
    var machine: ActionMachine

    init {
        this.machine = ActionMachine(actionType)
    }
}

class MachineState(var scene: String, var concernResult: Boolean      // 如果事件失败，则任务中断失败
                   , var delay: Long, var commandCode: Int) {
    var canSkip: Boolean = false
    var waitForContentChange: Boolean = false
    var obj: Any? = null
    var extraScene: Array<String>? = null         // 有可能有多个场景可执行相同的步骤

    constructor(scene: String, commandCodes: Int) : this(scene, false, commandCodes) {}

    constructor(scene: String, concernResult: Boolean, commandCodes: Int) : this(scene, concernResult, AccessibilityCommandHandler.DEFAULT_COMMAND_INTERVAL, commandCodes) {}

    private fun hasExtraSceneMatch(scene: String): Boolean {
        if (extraScene == null) return false
        for (s in extraScene!!) {
            if (s == scene) {
                return true
            }
        }
        return false
    }

    fun isSceneMatch(scene: String): Boolean {
        return this.scene == scene || hasExtraSceneMatch(scene)
    }
}
